### ISSUE TITLE
Equality of AuthState Objects

### ISSUE THREAD

**Issue by cakl** [openid/AppAuth-iOS#122]
Does AppAuth provide any functionality to compare or "check for equality" of ```OIDAuthState``` objects? AFAIK isEqual is not implemented. As AppAuth does not provide the serialization of the authstate, writing unit tests for the serialization layer seems hard without any possibility to check for equality.

**Comment by WilliamDenniss** [comment]
OIDAuthState does provide serialization through it's implementation of `NSSecureCoding`.  We just don't *store* the serialized object for you.

We have a unit test in the library to test  the serialization and deserialization (`OIDAuthStateTests.m:testSecureCoding`). Any interest to review this existing unit test and propose a PR if you see any issues, or missing coverage?

**Comment by StevenEWright** [comment]
@williamdenniss isn’t this a feature request to add ‘isEqual:’ and ‘hash’ overrides to all our model classes, not a question?

**Comment by WilliamDenniss** [comment]
Yes, but the premise ("AppAuth does not provide the serialization of the authstate") is wrong. I'm all for implementing isEqual, I just don't think it's needed for this use-case since we already cover it.

**Comment by StevenEWright** [comment]
I assumed the OP meant “persistence” not “serialization”. Sounded like they wanted to make sure that after they archive and persist the data that they can get it back later. To create a unit test for that flow you’d want isEqual:

Just guessing what the op meant - but that’s what I’d want too.

**Comment by WilliamDenniss** [comment]
Ah, fair enough. Well I like the idea of adding those methods, we should change the current tests to use those methods and move whatever we had to test isEqual…

**Comment by julienbodet** [comment]
Hi @StevenEWright I see you're assigned to this issue. I started working on this a couple weeks ago but didn't had time to finish. I already implemented the `isEqual` methods for the `OIDAuthState` object and all the child objects. I can share it with you by pushing it in a branch if it can helps you! Let me know.

**Comment by StevenEWright** [comment]
Oh! I have not started. If you’re keen to do it, that’s awesome!! If you don’t have plans on working on it in the next couple weeks, maybe I could make PRs to your branch and we could both contribute?

**Comment by julienbodet** [comment]
I'm not sure to have time to work on this in the coming weeks, so here is my [branch](https://github.com/julienbodet/AppAuth-iOS/tree/issue/122-equality). I would be happy to contribute with you on that one!

**Comment by StevenEWright** [comment]
Awesome!!! I’ll follow up on your branch tonight/tomorrow. Looks like you’ve already implemented a number of the isEqual: methods. We probably need to implement hash and add some unit tests, so I can send you a PR for those!

**Comment by julienbodet** [comment]
I implemented `isEqual:` method for all the objects implementing `NSSecureCoding` and the equality check is based only on the encoded properties.

**Comment by muthenberg** [comment]
Is there someone still working on this? Would be great to get this solution for testing.

**Comment by ywsang** [comment]
Hi there! Unfortunately this is not on our roadmap at this point in time. That being said, with the open source nature of the project, if you've implemented a solution that works, feel free to create a PR (and tag this issue) and we would love to review it!
